### PR TITLE
Fix #568: recognize provided expressions

### LINKED ISSUE #568: How to read the AST with provided types
I'm trying to compile a script using the JSON type provider with Fable. However, whenever I try to access an expression using a provided type an exception is thrown.

#### Repro steps

Try to compile the following script with Fable.

```fsharp
#r "../../../bin/FSharp.Data.dll"
open FSharp.Data
type Simple = JsonProvider<""" { "name":"John", "age":94 } """>
let simple = Simple.Parse(""" { "name":"Tomas", "age":4 } """)
simple.Age
```

#### Expected behavior

The AST is fully accessible and can be compiled to JS.

#### Actual behavior

When examining the tree and reaching `simple.Age`, trying to match the expression against any of the active patterns in `BasicPatterns` module results in the following error:

`A call to 'FSharp.Data.Runtime.JsonValueOptionAndPath::get_Path(...)' could not be resolved`

Stack trace:

```
  at Microsoft.FSharp.Compiler.SourceCodeServices.ExprUtilsImpl.bindILMethodRefToSymbol (Microsoft.FSharp.Compiler.SourceCodeServices.cenv cenv, range m, Microsoft.FSharp.Compiler.AbstractIL.ILMethodRef ilMethRef) <0xd6f8410 + 0x0073f> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExprConvert.ConvExprPrim (Microsoft.FSharp.Compiler.SourceCodeServices.cenv cenv, Microsoft.FSharp.Compiler.SourceCodeServices.ExprTranslationEnv env, Microsoft.FSharp.Compiler.Expr expr) <0xd6ee7c8 + 0x009bf> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExprConvert.ConvExprPrimLinear (Microsoft.FSharp.Compiler.SourceCodeServices.cenv cenv, Microsoft.FSharp.Compiler.SourceCodeServices.ExprTranslationEnv env, Microsoft.FSharp.Compiler.Expr expr, Microsoft.FSharp.Core.FSharpFunc`2 contf) <0xd6f65f0 + 0x00463> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExprConvert.ConvExprPrim (Microsoft.FSharp.Compiler.SourceCodeServices.cenv cenv, Microsoft.FSharp.Compiler.SourceCodeServices.ExprTranslationEnv env, Microsoft.FSharp.Compiler.Expr expr) <0xd6ee7c8 + 0x00a1f> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExprConvert.ConvExprPrimLinear (Microsoft.FSharp.Compiler.SourceCodeServices.cenv cenv, Microsoft.FSharp.Compiler.SourceCodeServices.ExprTranslationEnv env, Microsoft.FSharp.Compiler.Expr expr, Microsoft.FSharp.Core.FSharpFunc`2 contf) <0xd6f65f0 + 0x00463> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExprConvert+ConvExprOnDemand@903.Invoke (Microsoft.FSharp.Core.Unit unitVar0) <0xd6ee758 + 0x0002f> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExpr.get_E () <0xd6ee710 + 0x00030> in <filename unknown>:0 
  at Microsoft.FSharp.Compiler.SourceCodeServices.BasicPatterns.|Call|_| (Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExpr e) <0xe089fa0 + 0x00017> in <filename unknown>:0 
  at FSI_0002+printDecls@48-10.Invoke (Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExpr _arg1) <0xe089e38 + 0x00013> in <filename unknown>:0 
  at FSI_0002.lookup (Microsoft.FSharp.Core.FSharpFunc`2 f, Microsoft.FSharp.Compiler.SourceCodeServices.FSharpExpr expr) <0xe089da8 + 0x00018> in <filename unknown>:0 
  at FSI_0002+printDecls@36.Invoke (Int32 i, Microsoft.FSharp.Compiler.SourceCodeServices.FSharpImplementationFileDeclaration decl) <0xd6e6c28 + 0x0048b> in <filename unknown>:0 
  at Microsoft.FSharp.Collections.SeqModule.IterateIndexed[T] (Microsoft.FSharp.Core.FSharpFunc`2 action, IEnumerable`1 source) <0xd6e2180 + 0x000d0> in <filename unknown>:0 
  at FSI_0002.printDecls (System.String prefix, Microsoft.FSharp.Collections.FSharpList`1 decls) <0xd6e6a40 + 0x00067> in <filename unknown>:0
```

#### Known workarounds

None.

#### Related information 

* Operating system: OSX El Capitán
* Branch: FSharp.Compiler.Service 3.0.0
* .NET Runtime, CoreCLR or Mono Version: Mono 4.4.0
* Editing Tools (e.g. Visual Studio Version): VS Code/Xamarin

### PR DESCRIPTION
**Pull request by Jand42**
Expression API recognizes many calls generated by type providers that previously thrown an exception.

Changes included:
* Add a case `TyparRigidity.Unresolved` to notate type parameters with unknown type constraints. These are ignored when doing signature comparison checks
* An unscoping for looking up IL members, version information can be missing from the reference coming from a TP
* Better recursive stripping of type aliases and measures
* Added various tests, TestTP is now included in main solution

One TP test method `ChoiceConstructionAndMatch` fails during F# type checking, union generated subclass fails to resolve to F# symbol, this test is currently commented out

### CODE REVIEW

**Comment by Jand42** [comment]
Still not correct for use with FSharp.Data+WebSharper 4, adding more tests

**Comment by dsyme** [comment]
@Jand42 Wow, impressive

I need to look very closely at the changes in tast.fs and TastOps.fs - if there are any changes there that you think are unnecessary then taking them out would simplify accepting this

One compilation problem:

    [00:06:05] C:\projects\fsharp-compiler-service\src\fsharp\ConstraintSolver.fs(348,11): error FS0025: Incomplete pattern matches on this expression. For example, the value '(_,Unresolved)' may indicate a case not covered by the pattern(s).

**Comment by Jand42** [comment]
@dsyme Probably all changes in tast.fs and TastOps.fs can be reverted. I am adding more tests to gather and print all method signatures, not just the shape of the expression, I have already found some mistakes.

**Comment by Jand42** [comment]
@dsyme I have reverted every change to tast.fs and TastOps.fs, they were not needed.

Added a printer for method reference signatures, and tests with it.

One limitation remains, that if overloads are created with using the same name in `CompiledName`, it cannot be resolved.

**Comment by dsyme** [comment]
Oh that/s great, makes it much easier. 

There's still a failure in CI build

```
[00:08:34] C:\Program Files\dotnet\dotnet.exe compile-fsc @C:\projects\fsharp-compiler-service\tests\service\obj\Release\netcoreapp1.0\dotnet-compile.rsp returned Exit Code 1
[00:08:34] 
[00:08:34] C:\projects\fsharp-compiler-service\tests\service\ExprTests.fs(863,9): error FS0039: The namespace or module 'ProjectCracker' is not defined
[00:08:34] 
```

File: src/fsharp/vs/Exprs.fs [hunk 1]
```diff
@@ -752,6 +679,190 @@ module FSharpExprConvert =
...
+                    let mdef = resolveILMethodRefWithRescope unscopeILType tcref.ILTyconRawMetadata ilMethRef 
+                    let minfo = MethInfo.CreateILMeth(cenv.amap, m, enclosingType, mdef)                     
+                    FSharpMemberOrFunctionOrValue(cenv, minfo) |> makeCall |> Some
+                with _ -> 
+                    None
+            else
+                None
+
+        // Otherwise try to bind it to an F# symbol
+        match try1 with
```
Comment (dsyme) [inline src/fsharp/vs/Exprs.fs]: Could you extract this large expression out to a separate helper function please, outside the big ``let rec``? Along the lines of ``bindILMethodRefToSymbol`` in the previous code. Thanks

File: src/fsharp/vs/Exprs.fs [hunk 2]
```diff
@@ -752,6 +679,190 @@ module FSharpExprConvert =
...
+                let e = ilMethRef.EnclosingTypeRef
+                let parent = ILTypeRef.Create(e.Scope, e.Enclosing.Tail, e.Enclosing.Head)
+                Import.ImportILTypeRef cenv.amap m parent, Some e.Name
+                
+        let enclosingType = generalizedTyconRef tcref
+        
+        let makeCall minfo =
+            ConvObjectModelCallLinear cenv env (isNewObj, minfo, enclTypeArgs, methTypeArgs, callArgs) id   
+
+        let makeFSCall isMember (vr: ValRef) =
```
Comment (dsyme) [inline src/fsharp/vs/Exprs.fs]: Please extract ``makeFSCall`` and ``makeFSExpr`` out to helper functions in a module at the top of the file, independent of the big ``let rec``.  Along the lines of ``bindILMethodRefToSymbol`` in the previous code.  I'd like to see that they are `///` comments and be independent of the rest of the code.

**Comment by dsyme** [comment]
Still seeing the following fail for .NET Core: 

    [00:08:16] C:\projects\fsharp-compiler-service\tests\service\ExprTests.fs(865,9): error FS0039: The namespace or module 'ProjectCracker' is not defined

**Comment by Jand42** [comment]
I have disabled runnint TP tests on .NET core for now.

Added some tests and fixes for accessing module members and `get_Tag` of non-nullable unions.

I have looked into separating whole `ConvILCall` or just the inner helper functions, but it would just make it more convoluted with I think. The difference is that old implementation only returned an `FSharpMemberOrFunctionOrValue` and now it constructs different kinds of expressions, so it can't be on the top of the file without returning a union result which had to be matched to create the final expression. Also, `ConvLValueExpr` is used depending on the kind of F# member matched, otherwise an extra AddressOf would remain in the expression.

**Comment by dsyme** [comment]
OK. Great work.  Do you think this is now ready to merge?

**Comment by Jand42** [comment]
I will do some testing with it today on FSharp.Data+WebSharper 4, will report back.

**Comment by Jand42** [comment]
Everything worked ok (building full WebSharper 4 stack, testing try.websharper), I hope to release all very soon

**Comment by dsyme** [comment]
6.0.2 released. Awesome job fixing this!!!

**Pull request merged**